Gems/Atom/Tools/AtomToolsFramework/Code/Source/Document/AtomToolsDocumentMainWindow.cpp

```diff
@@ -150,29 +125,39 @@ namespace AtomToolsFramework
         });
 
         m_actionSaveAll = CreateActionAtPosition(m_menuFile, insertPostion, "Save A&ll", [this]() {
-            bool result = false;
-            AtomToolsDocumentSystemRequestBus::EventResult(
-                result, m_toolId, &AtomToolsDocumentSystemRequestBus::Events::SaveAllDocuments);
-            if (!result)
+
+            for (int index = 0; index < m_tabWidget->count(); index++)
             {
-                SetStatusError(tr("Document save all failed"));
+                if (!SaveDocument(GetDocumentTabId(index)))
+                {
+                    // Stop if there is any save failed or cancel
+                    break;
+                }
             }
         });
         m_menuFile->insertSeparator(insertPostion);
 
         m_actionClose = CreateActionAtPosition(m_menuFile, insertPostion, "&Close", [this]() {
             const AZ::Uuid documentId = GetCurrentDocumentId();
-            AtomToolsDocumentSystemRequestBus::Event(m_toolId, &AtomToolsDocumentSystemRequestBus::Events::CloseDocument, documentId);
+            if (CloseDocumentCheck(documentId))
+            {
+                AtomToolsDocumentSystemRequestBus::Event(m_toolId, &AtomToolsDocumentSystemRequestBus::Events::CloseDocument, documentId);
+            }
         }, QKeySequence::Close);
 
         m_actionCloseAll = CreateActionAtPosition(m_menuFile, insertPostion, "Close All", [this]() {
-            AtomToolsDocumentSystemRequestBus::Event(m_toolId, &AtomToolsDocumentSystemRequestBus::Events::CloseAllDocuments);
+            if (CloseAllDocumentCheck())
+            {
+                AtomToolsDocumentSystemRequestBus::Event(m_toolId, &AtomToolsDocumentSystemRequestBus::Events::CloseAllDocuments);
+            }
```

Comment:
I tried this before and there is a problem here.
Because `AtomToolsDocumentSystemRequestBus::Events::CloseDocument` actually closes the document, `m_tabWidget->count()` will change, and this will mess up the loop.
That's why I change it to check first, then close the tabs with respective functions.

Reply:
I see. That would just require a minor update to loop over the tabs and build a set of document IDs. Then loop over the document IDs and close them. You could even pass in the set of documents to close instead of passing in the document to ignore. Then the function would be used consistently for all close actions.


```
const AZStd::vector<AZ::Uuid> GetOpenDocumentIds() const
{
    AZStd::vector<AZ::Uuid> documentIds;
    documentIds.reserve(m_tabWidget->count());
    for (int index = 0; index < m_tabWidget->count(); index++)
    {
        documentIds.push_back(GetDocumentTabId(index));
    }
    return documentIds;
}

bool CloseDocuments(const AZStd::vector<AZ::Uuid>& documentIds)
{
    for (const auto& documentId : documentIds)
    {
        if (!CloseDocumentCheck(documentId))
        {
            return false;
        }

        bool result = false;
        AtomToolsDocumentSystemRequestBus::EventResult(
            result, m_toolId, &AtomToolsDocumentSystemRequestBus::Events::CloseDocument, documentId);
        if (!result)
        {
            return false;
        }
    }
    return true;
}

CloseDocuments(GetOpenDocumentIds());
```

Reply:
Thanks for the idea!